pyo3-macros-backend/src/params.rs

```diff
@@ -239,80 +216,96 @@ pub fn impl_arg_params(
                 #from_py_with
         },
         param_conversion,
-    ))
+    )
+}
+
+fn impl_arg_param(
+    arg: &FnArg<'_>,
+    pos: usize,
+    option_pos: &mut usize,
+    holders: &mut Holders,
+    ctx: &Ctx,
+) -> TokenStream {
+    let Ctx { pyo3_path } = ctx;
+    let args_array = syn::Ident::new("output", Span::call_site());
+
+    match arg.kind {
+        FnArgKind::Regular { .. } => {
+            let from_py_with = syn::Ident::new(&format!("from_py_with_{}", pos), Span::call_site());
+            let arg_value = quote!(#args_array[#option_pos].as_deref());
+            *option_pos += 1;
+            let tokens = impl_regular_arg_param(arg, from_py_with, arg_value, holders, ctx);
+            check_arg_for_gil_refs(tokens, holders.push_gil_refs_checker(arg.ty.span()), ctx)
+        }
+        FnArgKind::VarArgs => {
+            let holder = holders.push_holder(arg.name.span());
+            let name_str = arg.name.to_string();
+            quote_spanned! { arg.name.span() =>
+                #pyo3_path::impl_::extract_argument::extract_argument(
+                    &_args,
+                    &mut #holder,
+                    #name_str
+                )?
+            }
+        }
+        FnArgKind::KwArgs => {
+            let holder = holders.push_holder(arg.name.span());
+            let name_str = arg.name.to_string();
+            quote_spanned! { arg.name.span() =>
+                #pyo3_path::impl_::extract_argument::extract_optional_argument(
+                    _kwargs.as_deref(),
+                    &mut #holder,
+                    #name_str,
+                    || ::std::option::Option::None
+                )?
+            }
+        }
+        FnArgKind::Py => quote! { py },
+        FnArgKind::CancelHandle { .. } => quote! { __cancel_handle },
+    }
 }
 
 /// Re option_pos: The option slice doesn't contain the py: Python argument, so the argument
 /// index and the index in option diverge when using py: Python
-pub(crate) fn impl_arg_param(
+pub(crate) fn impl_regular_arg_param(
     arg: &FnArg<'_>,
     from_py_with: syn::Ident,
     arg_value: TokenStream, // expected type: Option<&'a Bound<'py, PyAny>>
     holders: &mut Holders,
     ctx: &Ctx,
-) -> Result<TokenStream> {
+) -> TokenStream {
+    // FIXME: Is there a better way to guarantee this in the type system?
```

Comment:
You could write
```rust
pub(crate) fn impl_regular_arg_param(
    arg: &FnArg<'_>,
    from_py_with: syn::Ident,
    arg_value: TokenStream, // expected type: Option<&'a Bound<'py, PyAny>>
    holders: &mut Holders,
    default: Whatever,
    ty_opt: Whatever,
    ctx: &Ctx,
) -> TokenStream {
    let FnArg {
        name,
        ty,
        attrs,
        ..
    } = arg;
```

or you could make enum variant types:

```rust
pub struct Regular<'a> {
    pub name: &'a syn::Ident,
    pub ty: &'a syn::Type,
    pub attrs: PyFunctionArgPyO3Attributes,
    pub kind: FnArgKind<'a>,
    default: Option<syn::Expr>,
    ty_opt: Option<&'a syn::Type>,
}

pub struct VarArgs<'a> {
    pub name: &'a syn::Ident,
    pub ty: &'a syn::Type,
    pub attrs: PyFunctionArgPyO3Attributes,
    pub kind: FnArgKind<'a>,
}

pub enum FnArg<'a> {
    Regular(Regular<'a>)
    VarArgs(VarArgs<'a>),
    ...
}
```

Reply:
Thanks for the feedback 👍, I though about using enum variant types, but initially decided against them because it might get very verbose. But maybe it is beneficial to have a separate type for each kind of argument. I think I'll give it a try and see how it turns out.